# Tarea 1 Bases de Datos 2023-1 Grupo 31


## Integrantes:
* Agustín Aravena - 202173022-2
* Sofía Meléndez - 202173041-9

## Requisitos para Correr el Notebook

Para poder utilizar este Notebook se debe tener instalado:
- **Python 3**
- **Jupyter**
- La librería **ipython-sql** (Poder ejecutar sql en Jupyter Notebook)
- PostgreSQL y su conector desde Python(**psycopg2**)

**IMPORTANTE:** Este Jupyter Notebook se conectará a su servicio de Base de Datos PostgreSQL de manera **local**, por lo que **no funcionará correctamente en el entorno Google Collab.**

Para instalar las dependencias, ejecute (**recuerde actualizar pip3**):

In [ ]:
! pip3 install ipython-sql
! pip3 install psycopg2

#Use to generate names
! pip3 install names

Defaulting to user installation because normal site-packages is not writeable



[notice] A new release of pip available: 22.3.1 -> 23.0.1
[notice] To update, run: python.exe -m pip install --upgrade pip


Defaulting to user installation because normal site-packages is not writeable



[notice] A new release of pip available: 22.3.1 -> 23.0.1
[notice] To update, run: python.exe -m pip install --upgrade pip


Defaulting to user installation because normal site-packages is not writeable



[notice] A new release of pip available: 22.3.1 -> 23.0.1
[notice] To update, run: python.exe -m pip install --upgrade pip


## Importar dependencias

In [1]:
from datetime import datetime
import random as r
import names

#Librerías para PostgreSQL
import psycopg2

Para conectarse a la base de datos, completen los datos relevantes en la variable DATABASE_URL.

* Cambie `user` por el usuario que utilizó en Postgres para acceder a la Base de Datos. Por defecto `postgres`.
* Cambie `password` por la contraseña de dicho usuario. **Recuerde borrarla antes de entregar su tarea.**
* Cambie `port` por el puerto en que se ejecuta su servicio de Base de Datos. Por defecto `5432`.
* Cambie `database` por el nombre de la BD (Schema) que creó en Postgres.

Ejemplo: `postgresql://postgres:pass_segura@localhost:5432/tarea-bd`

In [2]:
%reload_ext sql
%env DATABASE_URL=postgresql://postgres:password@localhost:5432/dbname

env: DATABASE_URL=postgresql://postgres:Hogwartsismyhome.07@localhost:5432/tarea1


Para conectarse de forma declarativa a la Base de Datos, utilice el siguiente código:

In [3]:
#Recuerde cambiar los valores de USER, DBNAME y PASSWORD por los mismos de la celda superior.
connection = psycopg2.connect(
   host="localhost",
   user="postgres",
   dbname='dbname',
   password="password"
)

print(connection)
cursor = connection.cursor()

<connection object at 0x00000269840D2F00; dsn: 'user=postgres password=xxx dbname=tarea1 host=localhost', closed: 0>


## Crear tablas (36 pts.)

Aquí inserta el código SQL que creara las tablas de la base de datos:

### Tabla "facciones" (6 pts)

In [4]:
%%sql
CREATE TABLE facciones (
    id INTEGER GENERATED ALWAYS AS IDENTITY,
    nombre VARCHAR(45) NOT NULL,
    descripcion VARCHAR(45) NOT NULL,
    PRIMARY KEY (id)
);

(psycopg2.errors.DuplicateTable) la relación «facciones» ya existe

[SQL: CREATE TABLE facciones (
    id INTEGER GENERATED ALWAYS AS IDENTITY,
    nombre VARCHAR(45) NOT NULL,
    descripcion VARCHAR(45) NOT NULL,
    PRIMARY KEY (id)
);]
(Background on this error at: https://sqlalche.me/e/20/f405)


### Tabla "trabajos" (6 pts)

In [5]:
%%sql
CREATE TABLE trabajos(
    id INTEGER GENERATED ALWAYS AS IDENTITY,
    nombre VARCHAR(45) NOT NULL,
    descripcion VARCHAR(45) NOT NULL,
    sueldo INTEGER NOT NULL,
    PRIMARY KEY (id)
);

 * postgresql://postgres:***@localhost:5432/tarea1
(psycopg2.errors.DuplicateTable) la relación «trabajos» ya existe

[SQL: CREATE TABLE trabajos(
    id INTEGER GENERATED ALWAYS AS IDENTITY,
    nombre VARCHAR(45) NOT NULL,
    descripcion VARCHAR(45) NOT NULL,
    sueldo INTEGER NOT NULL,
    PRIMARY KEY (id)
);]
(Background on this error at: https://sqlalche.me/e/20/f405)


### Tabla "personas" (6 pts)

In [6]:
%%sql
CREATE TABLE personas (
    id INTEGER GENERATED ALWAYS AS IDENTITY,
    nombre VARCHAR(45) NOT NULL,
    apellidos VARCHAR(45) NOT NULL,
    fecha_nacimiento DATE NOT NULL,
    infectado BOOLEAN NOT NULL,
    conyugue INTEGER,
    PRIMARY KEY (id),
    FOREIGN KEY (conyugue) REFERENCES personas(id)
);

 * postgresql://postgres:***@localhost:5432/tarea1
(psycopg2.errors.DuplicateTable) la relación «personas» ya existe

[SQL: CREATE TABLE personas (
    id INTEGER GENERATED ALWAYS AS IDENTITY,
    nombre VARCHAR(45) NOT NULL,
    apellidos VARCHAR(45) NOT NULL,
    fecha_nacimiento DATE NOT NULL,
    infectado BOOLEAN NOT NULL,
    conyugue INTEGER,
    PRIMARY KEY (id),
    FOREIGN KEY (conyugue) REFERENCES personas(id)
);]
(Background on this error at: https://sqlalche.me/e/20/f405)


### Tabla "cuentas_bancarias" (6 pts)

In [9]:
%%sql
CREATE TABLE cuentas_bancarias (
    id INTEGER GENERATED ALWAYS AS IDENTITY,
    tipo_de_cuenta VARCHAR(45) NOT NULL,
    banco VARCHAR(45) NOT NULL,
    saldo INTEGER NOT NULL,
    id_persona INTEGER NOT NULL,
    PRIMARY KEY (id),
    FOREIGN KEY (id_persona) REFERENCES personas (id)
);

 * postgresql://postgres:***@localhost:5432/tarea1
(psycopg2.errors.DuplicateTable) la relación «cuentas_bancarias» ya existe

[SQL: CREATE TABLE cuentas_bancarias (
    id INTEGER GENERATED ALWAYS AS IDENTITY,
    tipo_de_cuenta VARCHAR(45) NOT NULL,
    banco VARCHAR(45) NOT NULL,
    saldo INTEGER NOT NULL,
    id_persona INTEGER NOT NULL,
    PRIMARY KEY (id),
    FOREIGN KEY (id_persona) REFERENCES personas (id)
);]
(Background on this error at: https://sqlalche.me/e/20/f405)


### Tabla "persona_pertenece_faccion" (6 pts)

In [12]:
%%sql
CREATE TABLE persona_pertenece_faccion(
    id_persona INTEGER,
    id_faccion INTEGER,
    CONSTRAINT persona_pertenece_faccion_PK PRIMARY KEY (id_persona,id_faccion),
    CONSTRAINT persona_pertenece_faccion_FK1 FOREIGN KEY (id_persona) REFERENCES personas(id),
    CONSTRAINT persona_pertenece_faccion_FK2 FOREIGN KEY (id_faccion) REFERENCES facciones(id)
    );

 * postgresql://postgres:***@localhost:5432/tarea1
Done.


[]

### Tabla "persona_tiene_trabajo" (6 pts)

In [13]:
%%sql
CREATE TABLE persona_tiene_trabajo(
    id_persona INTEGER,
    id_trabajo INTEGER,
    estado BOOLEAN NOT NULL,
    ultima_vez_realizado TIMESTAMP NOT NULL,
    CONSTRAINT persona_tiene_trabajo_PK PRIMARY KEY (id_persona,id_trabajo),
    CONSTRAINT persona_tiene_trabajo_FK1 FOREIGN KEY (id_persona) REFERENCES personas(id),
    CONSTRAINT persona_tiene_trabajo_FK2 FOREIGN KEY (id_trabajo) REFERENCES trabajos(id)
);

 * postgresql://postgres:***@localhost:5432/tarea1
Done.


[]

## Código Generador de Datos

Ejecute este código para cargar datos a las tablas creadas.

PD: Si el codigo tira error por "violar la unicidad de las PK" es por una coincidencia de numeros random, solamente volver a ejecutar el codigo

In [14]:
#Recuerde haber ejecutado todas las celdas anteriores!

# Generador de timestamp aleatorio.
def generate_timestamp_nacimiento():
    year = r.randint(1970,2022)
    month = r.randint(1,12)
    day = r.randint(1,31)
    
    hour = r.randint(0,23)
    minute = r.randint(0,59)
    second = r.randint(0,59)
    
    if month in [4, 6, 9, 11] and day > 30:
        day = 30
        
    if month == 2 and day > 28:
        day = 28
        
    return datetime(year, month, day, hour, minute, second)

def generate_timestamp_trabajo():
    year = r.randint(2019,2022)
    month = r.randint(1,12)
    day = r.randint(1,31)
    
    hour = r.randint(0,23)
    minute = r.randint(0,59)
    second = r.randint(0,59)
    
    if month in [4, 6, 9, 11] and day > 30:
        day = 30
        
    if month == 2 and day > 28:
        day = 28
        
    return datetime(year, month, day, hour, minute, second)

print("Borrando Datos Antiguos...")

# Limpia la base de datos.
connection.commit() # En caso de algun error
query = 'TRUNCATE TABLE "facciones", "trabajos", "personas", "cuentas_bancarias", "persona_pertenece_faccion", "persona_tiene_trabajo" RESTART IDENTITY'
cursor.execute(query)
connection.commit()

print("Generando Datos...")

facciones = [("FEDRA", "La Agencia Federal de Respuesta a Desastres"), ("Luciérnagas", "Grupo de milicias revolucionario"), ("Carroñeros", "Sin identificación"), ("Pythonia", "Grupo defensor de la libertad"), ("MaCdonia", "Los exiliados de la carceles"), ("JavaScriptia", "Grupo defensor de la paz")]
trabajos = [("Limpieza", "Encargado de limpiar las calles"),
("Limpiador de cloacas","Encargado de limpiar las cloacas"),
("Militar","Encargado de proteger la ciudad"),
("Explorador","Encargado de explorar la ciudad"),
("Médico","encargado de curar a los enfermos"),
("Profesor","Encargado de enseñar a los niños"),
("Cremador","Encargado de quemar los cadáveres"),
("Científico","Encargado de investigar"),
("Cocinero","Encargado de cocinar"),
("Carpintero","Encargado de construir"),
("Cerrajero","Encargado de reparar las cerraduras"),
("Cartero","Encargado de repartir los correos"),
("Cajero","Encargado de cobrar los impuestos")]
tipos_de_cuenta_bancaria = ["Cuenta Vista", "Cuenta Ahorro", "Cuenta Corriente", "Cuenta RUT", "Cuenta de Inversion", "Cuenta PrePago", "Cuenta de Credito"]
bancos = ["FedraBank S.A", "Banco Pedrito Pascal SpA", "Banco Pythonia S.A", "Banco de la Nación S.A", "Banco Santander SpA", "Banco BCI SpA", "Banco Estado SpA", "Banco Consorcio S.A", "Banco BBVA", "Banco Scotiabank"]

# Variables.
n_facciones = len(facciones)
n_trabajos = len(trabajos)
n_personas = 1000
n_cuentas_bancarias = 120
n_persona_pertenece_faccion = 300
n_persona_tiene_trabajo = 300
n_casados = 100

# Genera personas.
personas = []
for _ in range(n_personas):
    nombre = names.get_first_name()
    apellidos = names.get_last_name()+" "+names.get_last_name()
    fecha_nacimiento = generate_timestamp_nacimiento()
    infectado = r.choice([True, False])
    
    personas.append((nombre, apellidos, fecha_nacimiento, infectado, None))


# Genera cuentas bancarias.
cuentas_bancarias = []
for _ in range(n_cuentas_bancarias):
    tipo_de_cuenta = r.choice(tipos_de_cuenta_bancaria)
    banco = r.choice(bancos)
    saldo = r.randint(0, 10000)
    id_persona = r.choice(range(1, n_personas+1))
    
    cuentas_bancarias.append((tipo_de_cuenta, banco, saldo, id_persona))
    
    
# Generate persona_pertenece_faccion.
persona_pertenece_faccion = set([(r.choice(range(1, n_personas+1)), r.choice(range(1, n_facciones+1))) for _ in range(1, n_persona_pertenece_faccion+1)])

# Generate persona_tiene_trabajo.
persona_tiene_trabajo_prev = set([(r.choice(range(1, n_personas+1)), r.choice(range(1, n_trabajos+1))) for _ in range(1, n_persona_tiene_trabajo+1)])
persona_tiene_trabajo = [(id_persona, id_trabajo, r.choice([True, False]), generate_timestamp_trabajo()) for id_persona, id_trabajo in persona_tiene_trabajo_prev]
    
# SQL Statements.
inserts = [
    'INSERT INTO "facciones" ("nombre", "descripcion") VALUES (%s, %s);',
    'INSERT INTO "trabajos" ("nombre", "descripcion", "sueldo") VALUES (%s, %s, %s);',
    'INSERT INTO "personas" ("nombre", "apellidos", "fecha_nacimiento", "infectado", "conyugue") VALUES (%s, %s, %s, %s, %s);',
    'INSERT INTO "cuentas_bancarias" ("tipo_de_cuenta", "banco", "saldo", "id_persona") VALUES (%s, %s, %s, %s);',
    'INSERT INTO "persona_pertenece_faccion" ("id_persona", "id_faccion") VALUES (%s, %s);',
    'INSERT INTO "persona_tiene_trabajo" ("id_persona", "id_trabajo", "estado", "ultima_vez_realizado") VALUES (%s, %s, %s, %s);'
]

for faccion in facciones:
    cursor.execute(inserts[0], faccion)

for trabajo in trabajos:
    t = (trabajo[0], trabajo[1], r.randint(100, 5000))
    cursor.execute(inserts[1], t)

# Crear una lista con personas casadas y otra con personas solteras. donde se cambie la ide de la persona que se casó.

for persona in personas:
    cursor.execute(inserts[2], persona)
    
for cb in cuentas_bancarias:
    cursor.execute(inserts[3], cb)

for ppf in persona_pertenece_faccion:
    cursor.execute(inserts[4], ppf)
    
for ptt in persona_tiene_trabajo:
    cursor.execute(inserts[5], ptt)
    
connection.commit()


# Generar Matrimonios
n_repetidos = []

for i in range(n_casados):
    random = r.randint(1, n_personas)
    if random in n_repetidos:
        continue
    else:
        random2 = r.randint(1, n_personas)
        if random2 in n_repetidos:
            continue
        else:
            n_repetidos.append(random)
            n_repetidos.append(random2)
            cursor.execute("UPDATE personas SET conyugue = %s WHERE id = %s", (random2, random))
            cursor.execute("UPDATE personas SET conyugue = %s WHERE id = %s", (random, random2))
            connection.commit()

print("Todo Listo!")

Borrando Datos Antiguos...
Generando Datos...
Todo Listo!


## Consultas (64 pts.)

Escriba aquí las consultas pedidas. Recuerde agregar una descripción _breve_ de cómo funciona la solución. **Escriba las consultas de forma ordenada.**

**NOTA:** Cuando en una consulta diga "X cosa" significa que ustedes deben de elegir ese X por lo que para cualquier otro X debe de funcionar de igual forma.

### Consulta 1 (6 pts)
**ID, nombre y primer apellido de los habitantes de FEDRA**

Ejemplo:

|ID |  Nombre      |  Apellido  |
|---|-----------|-----------|
| 12 |Joel | Miller |
|21 | Ellie | Williams |
| ... | ... | ... |

**Solución:** Seleccionamos el id, el nombre y los apellidos de la tabla persona. Para poder tener solo el primer apellido usamos la funcion left() y la funcion position(), la funcion position es para que cuando identifique un espacio lo que le sigue no se agregue a la columna apellido.


In [204]:
%%sql

/* Reemplace estas queries por las que resuelven las consultas pedidas. */

SELECT
    id AS "ID", 
    nombre AS "Nombre", 
    left(apellidos, position(' ' in apellidos)) AS "Apellido"
FROM 
    personas
ORDER BY 
    id ASC

 * postgresql://postgres:***@localhost:5432/tarea1
1000 rows affected.


[(1, 'Myrtis', 'Lynum '),
 (2, 'Heather', 'Nolin '),
 (3, 'Shane', 'Rozanski '),
 (4, 'Victoria', 'Stafford '),
 (5, 'Hazel', 'Hartel '),
 (6, 'Justin', 'Carney '),
 (7, 'Michell', 'Gonser '),
 (8, 'Teresa', 'Session '),
 (9, 'Elizabeth', 'Addams '),
 (10, 'Julian', 'Kay '),
 (11, 'Shelly', 'Myott '),
 (12, 'Maria', 'Piccinone '),
 (13, 'Myrtle', 'Mcclaskey '),
 (14, 'Joann', 'Moran '),
 (15, 'August', 'Smith '),
 (16, 'Nicole', 'Tartaglione '),
 (17, 'Phillip', 'Newcomb '),
 (18, 'Sheri', 'Testa '),
 (19, 'Elizabeth', 'Howell '),
 (20, 'Maria', 'Hartman '),
 (21, 'William', 'Gardner '),
 (22, 'Raymond', 'Raybuck '),
 (23, 'Philip', 'Montoya '),
 (24, 'Brenda', 'Basham '),
 (25, 'Katherine', 'Beaudoin '),
 (26, 'Paul', 'Newsome '),
 (27, 'Joseph', 'Skaggs '),
 (28, 'Benjamin', 'Mcnally '),
 (29, 'Thomas', 'Patterson '),
 (30, 'Leroy', 'Frederick '),
 (31, 'Ramon', 'Wilhelm '),
 (32, 'Dolores', 'Randall '),
 (33, 'Suzanne', 'Babb '),
 (34, 'Lisa', 'Schmalz '),
 (35, 'Neal', 'Force '),
 (36, 'Dina', 'Turner '),
 (37, 'Robert', 'Garza '),
 (38, 'William', 'Starr '),
 (39, 'Christopher', 'Babbitt '),
 (40, 'Katrina', 'Shorey '),
 (41, 'Edward', 'Lowe '),
 (42, 'Christine', 'Beltran '),
 (43, 'Randi', 'Navarro '),
 (44, 'Carolyn', 'Kurz '),
 (45, 'Mary', 'Peacock '),
 (46, 'Rae', 'Reichman '),
 (47, 'Hugh', 'Mintz '),
 (48, 'Lizzie', 'Nicholson '),
 (49, 'Patricia', 'Mansfield '),
 (50, 'Amy', 'Mull '),
 (51, 'Mary', 'Potter '),
 (52, 'Clarence', 'Carrithers '),
 (53, 'Vincent', 'Jones '),
 (54, 'James', 'Hill '),
 (55, 'Raul', 'Richardson '),
 (56, 'Wallace', 'Baker '),
 (57, 'Mark', 'Thomas '),
 (58, 'Jordan', 'Edwards '),
 (59, 'Jennifer', 'Wilson '),
 (60, 'Tommy', 'Noonan '),
 (61, 'Helen', 'Tademy '),
 (62, 'Herman', 'Inoue '),
 (63, 'Helen', 'Donohoe '),
 (64, 'Norman', 'Cowger '),
 (65, 'Carlos', 'Mckinney '),
 (66, 'Cynthia', 'Minteer '),
 (67, 'William', 'Carnahan '),
 (68, 'Juan', 'Sheehy '),
 (69, 'Stanley', 'Shannon '),
 (70, 'Melissa', 'Spiney '),
 (71, 'James', 'Gaylor '),
 (72, 'Ronald', 'Kunz '),
 (73, 'Laura', 'Connors '),
 (74, 'Marvin', 'Hardy '),
 (75, 'Kenneth', 'Goldstein '),
 (76, 'Linda', 'Angeles '),
 (77, 'Penny', 'Warriner '),
 (78, 'Joseph', 'Hernandez '),
 (79, 'Logan', 'Widner '),
 (80, 'Mary', 'Jara '),
 (81, 'Maria', 'Petrulis '),
 (82, 'Terrance', 'Terepka '),
 (83, 'June', 'Parks '),
 (84, 'Chris', 'Muir '),
 (85, 'Annie', 'Bonner '),
 (86, 'Mary', 'Sherrod '),
 (87, 'Aaron', 'Sundt '),
 (88, 'Stuart', 'Ackerley '),
 (89, 'Barbara', 'Ling '),
 (90, 'Arthur', 'Mercurio '),
 (91, 'Kevin', 'Jones '),
 (92, 'Susie', 'Childs '),
 (93, 'Larry', 'Barrie '),
 (94, 'Maxine', 'Picou '),
 (95, 'Robyn', 'Guy '),
 (96, 'Edith', 'Stump '),
 (97, 'Lucille', 'Mcdonald '),
 (98, 'Barbara', 'Archer '),
 (99, 'Andrea', 'Shaffer '),
 (100, 'Michael', 'Tamayo '),
 (101, 'Tammy', 'Baldwin '),
 (102, 'Carol', 'Eugene '),
 (103, 'Michael', 'Brissett '),
 (104, 'Lauren', 'Benjamin '),
 (105, 'Melissa', 'Barry '),
 (106, 'Lottie', 'Munez '),
 (107, 'John', 'Devine '),
 (108, 'Dustin', 'Crawford '),
 (109, 'Mitchell', 'Briney '),
 (110, 'Roger', 'Gebhart '),
 (111, 'Robert', 'Hoskins '),
 (112, 'Guy', 'Merritt '),
 (113, 'Velma', 'Pruchnik '),
 (114, 'James', 'Saunders '),
 (115, 'Rachelle', 'Emme '),
 (116, 'Cody', 'Fox '),
 (117, 'Edward', 'Tennant '),
 (118, 'Melanie', 'Miley '),
 (119, 'Ronald', 'Sloan '),
 (120, 'Sara', 'Middleton '),
 (121, 'Barbara', 'Mackinnon '),
 (122, 'Nellie', 'Bonds '),
 (123, 'Carlos', 'Breedlove '),
 (124, 'William', 'Hyatt '),
 (125, 'Christopher', 'Roberts '),
 (126, 'Catherine', 'Richmond '),
 (127, 'Latisha', 'Hawkins '),
 (128, 'Lynn', 'Ashmore '),
 (129, 'Catherine', 'Bradley '),
 (130, 'Leslie', 'Hendricks '),
 (131, 'James', 'Laplante '),
 (132, 'Gary', 'Robinson '),
 (133, 'Paul', 'Campbell '),
 (134, 'Howard', 'Vasquez '),
 (135, 'Carol', 'Keeling '),
 (136, 'Georgia', 'Burton '),
 (137, 'Henry', 'Johnson '),
 (138, 'Edward', 'Byrd '),
 (139, 'Richard', 'Samayoa '),
 (140, 'Keith', 

### Consulta 2 (6 pts)
**Cantidad de personas nacidas después del inicio de la pandemia.**(inicio de la pandemia 01-12-2019)

Ejemplo:


| Cantidad     | 
|--------------|
| 292 | 

**Solución:** Utilizamos el condicional where para que contara a las personas nacidas después de la pandemia, y dado que nacieron luego de que iniciara la pandemia, el día 01-12-2019 no contaría, solamente contarían los que nacieron luego de ese día.

In [203]:
%%sql

/* Reemplace estas queries por las que resuelven las consultas pedidas. */

SELECT
     count(fecha_nacimiento) AS "Cantidad"
FROM 
    personas
WHERE 
    fecha_nacimiento > '01-12-2019'

 * postgresql://postgres:***@localhost:5432/tarea1
1 rows affected.


[(56,)]

### Consulta 3 (6 pts)
**ID, nombre, apellidos y saldo total de todas las personas**( Limitar a 20 filas para visualizar)

Ejemplo:

| ID    | Nombre | Apellido | Dinero | 
|----------------|--------|-|-|
| 13 | Tommy | Miller | 12501 |
| ... | ... | ... | ... |

**Solución:** Usamos un INNER JOIN para poder 'unir' la información de la tabla persona y de la tabla cuentas_bancarias.

In [202]:
%%sql

/* Reemplace estas queries por las que resuelven las consultas pedidas. */

SELECT
    personas.id AS "ID",
    personas.nombre AS "Nombre",
    left(apellidos, position(' ' in apellidos)) AS "Apellido",
    cuentas_bancarias.saldo AS "Dinero"
FROM 
    personas INNER JOIN cuentas_bancarias ON personas.id = cuentas_bancarias.id_persona
LIMIT 
    20

 * postgresql://postgres:***@localhost:5432/tarea1
20 rows affected.


[(4, 'Victoria', 'Stafford ', 2173),
 (10, 'Julian', 'Kay ', 4548),
 (17, 'Phillip', 'Newcomb ', 263),
 (34, 'Lisa', 'Schmalz ', 8841),
 (38, 'William', 'Starr ', 5452),
 (60, 'Tommy', 'Noonan ', 2117),
 (75, 'Kenneth', 'Goldstein ', 8755),
 (81, 'Maria', 'Petrulis ', 7035),
 (94, 'Maxine', 'Picou ', 8159),
 (105, 'Melissa', 'Barry ', 395),
 (110, 'Roger', 'Gebhart ', 7333),
 (132, 'Gary', 'Robinson ', 7470),
 (135, 'Carol', 'Keeling ', 2132),
 (180, 'Rosa', 'Shelton ', 8742),
 (190, 'Mary', 'Harris ', 5370),
 (208, 'Raul', 'Rodriquez ', 5251),
 (211, 'Rachel', 'Freeman ', 2284),
 (216, 'Elizabeth', 'Limmel ', 2558),
 (222, 'Cyril', 'Modica ', 8167),
 (248, 'Derek', 'Emmons ', 3929)]

### Consulta 4 (6 pts)
**Nombre de la Facción y la cantidad de trabajadores activos hasta fecha X.**

Ejemplo:

| Facción     | Cantidad de trabajadores |
|----------------|--------|
| FEDRA | 1273 |
| Luciérnagas | 121 |
| Carroñeros  | 72 |
| ...  | ... |

**Solución:** Usamos INNER JOIN para crear una integración de información de los trabajadores activos y su facción, luego realizamos un COUNT() para contar el total.

In [175]:
%%sql

/* Reemplace estas queries por las que resuelven las consultas pedidas. */

SELECT 
    facciones.nombre AS "Facción",
    COUNT(facciones.nombre) AS "Cantidad de trabajadores"
FROM 
    facciones
    INNER JOIN persona_pertenece_faccion ON facciones.id = persona_pertenece_faccion.id_faccion
    INNER JOIN personas ON persona_pertenece_faccion.id_persona = personas.id
    INNER JOIN persona_tiene_trabajo ON personas.id = persona_tiene_trabajo.id_persona WHERE estado = True 
GROUP BY 
    facciones.nombre


 * postgresql://postgres:***@localhost:5432/tarea1
6 rows affected.


[('JavaScriptia', 9),
 ('Pythonia', 6),
 ('MaCdonia', 11),
 ('FEDRA', 10),
 ('Luciérnagas', 7),
 ('Carroñeros', 5)]

### Consulta 5 (6 pts)
**Nombre de la Facción y su cantidad de personas infectadas.**

Ejemplo:

|  Facción  | Personas infectadas | 
|--------|-------|
| FEDRA | 251 |
| Luciérnagas | 17 |
| Carroñeros  | 12 |
| ...  | ... |

**Solución:** Hicimos INNER JOIN entre la tabla de las personas, de las facciones y de las personas_pertenecen_faccion para rescatar si esta infectado o no y su faccion correspondiente. Luego agrupamos según faccion.

In [87]:
%%sql

/* Reemplace estas queries por las que resuelven las consultas pedidas. */

SELECT
    facciones.nombre AS Faccion,
    COUNT (*) AS personas_infectadas
FROM 
    facciones INNER JOIN persona_pertenece_faccion ON facciones.id = persona_pertenece_faccion.id_faccion
    INNER JOIN personas ON persona_pertenece_faccion.id_persona = personas.id WHERE infectado = True
GROUP BY (facciones.nombre, personas.infectado)




 * postgresql://postgres:***@localhost:5432/tarea1
6 rows affected.


[('Carroñeros', 22),
 ('Pythonia', 14),
 ('Luciérnagas', 24),
 ('FEDRA', 25),
 ('JavaScriptia', 15),
 ('MaCdonia', 30)]

### Consulta 6 (6 pts)
**Nombre de la Facción y su trabajo más realizado.**

Ejemplo:

|  Facción  | Trabajo | 
|--------|-------|
| FEDRA | Limpiador de cloacas |
| Luciérnagas | Médico |
| Carroñeros  | Explorador |
| ...  | ... |

**Solución:** _Escriba aquí la descripción de su solución._

In [197]:
%%sql

/* Reemplace estas queries por las que resuelven las consultas pedidas. */
SELECT
    facciones.nombre AS faccion,
    trabajos.nombre,
    COUNT(trabajos.nombre)
FROM 
    facciones INNER JOIN persona_pertenece_faccion ON facciones.id = persona_pertenece_faccion.id_faccion
    INNER JOIN personas ON persona_pertenece_faccion.id_persona = personas.id
    INNER JOIN persona_tiene_trabajo ON personas.id = persona_tiene_trabajo.id_persona
    INNER JOIN trabajos ON persona_tiene_trabajo.id_trabajo = trabajos.id
GROUP BY 
    facciones.nombre, trabajos.nombre


 * postgresql://postgres:***@localhost:5432/tarea1
49 rows affected.


[('FEDRA', 'Limpieza', 1),
 ('Luciérnagas', 'Cerrajero', 1),
 ('Luciérnagas', 'Militar', 1),
 ('JavaScriptia', 'Científico', 3),
 ('Luciérnagas', 'Médico', 1),
 ('MaCdonia', 'Cajero', 5),
 ('Pythonia', 'Médico', 1),
 ('Luciérnagas', 'Carpintero', 2),
 ('Pythonia', 'Carpintero', 3),
 ('JavaScriptia', 'Explorador', 2),
 ('MaCdonia', 'Cerrajero', 3),
 ('FEDRA', 'Limpiador de cloacas', 1),
 ('Carroñeros', 'Limpieza', 1),
 ('Carroñeros', 'Científico', 1),
 ('Pythonia', 'Cajero', 2),
 ('Luciérnagas', 'Científico', 2),
 ('MaCdonia', 'Médico', 1),
 ('JavaScriptia', 'Cajero', 1),
 ('FEDRA', 'Cremador', 1),
 ('FEDRA', 'Cartero', 1),
 ('Pythonia', 'Limpiador de cloacas', 1),
 ('FEDRA', 'Militar', 2),
 ('Pythonia', 'Profesor', 1),
 ('FEDRA', 'Explorador', 3),
 ('FEDRA', 'Cerrajero', 3),
 ('MaCdonia', 'Limpiador de cloacas', 1),
 ('JavaScriptia', 'Médico', 1),
 ('FEDRA', 'Científico', 2),
 ('Carroñeros', 'Limpiador de cloacas', 3),
 ('FEDRA', 'Cajero', 2),
 ('Carroñeros', 'Cajero', 1),
 ('Luciérnagas', 'Cajero', 2),
 ('MaCdonia', 'Carpintero', 2),
 ('MaCdonia', 'Cartero', 1),
 ('MaCdonia', 'Explorador', 2),
 ('Luciérnagas', 'Limpiador de cloacas', 1),
 ('JavaScriptia', 'Limpieza', 1),
 ('Pythonia', 'Militar', 1),
 ('Carroñeros', 'Carpintero', 1),
 ('FEDRA', 'Carpintero', 1),
 ('JavaScriptia', 'Militar', 2),
 ('Carroñeros', 'Cerrajero', 3),
 ('JavaScriptia', 'Cerrajero', 1),
 ('Luciérnagas', 'Cocinero', 2),
 ('JavaScriptia', 'Carpintero', 1),
 ('Carroñeros', 'Militar', 1),
 ('FEDRA', 'Profesor', 1),
 ('Luciérnagas', 'Cremador', 1),
 ('MaCdonia', 'Profesor', 1)]

### Consulta 7 (8 pts)
**Nombre de la Facción y su Saldo total** (Es decir, la suma total de los saldos de cada persona perteneciente a la facción.)

Ejemplo:

|Facción | Saldo |
|--------|-|
| FEDRA | 523000 |
| Luciérnagas | 130023 |
| Carroñeros  | 17432 |
| ...  | ...|

**Solución:** _Escriba aquí la descripción de su solución._

In [156]:
%%sql

/* Reemplace estas queries por las que resuelven las consultas pedidas. */

SELECT
    facciones.nombre AS faccion,
    SUM(cuentas_bancarias.saldo) AS saldo
FROM 
    facciones INNER JOIN persona_pertenece_faccion ON facciones.id = persona_pertenece_faccion.id_faccion
    INNER JOIN personas ON persona_pertenece_faccion.id_persona = personas.id 
    INNER JOIN cuentas_bancarias ON personas.id = cuentas_bancarias.id_persona
GROUP BY
    facciones.nombre, saldo


 * postgresql://postgres:***@localhost:5432/tarea1
31 rows affected.


[('FEDRA', 6922),
 ('Pythonia', 2672),
 ('FEDRA', 7035),
 ('Pythonia', 4967),
 ('FEDRA', 3406),
 ('FEDRA', 5242),
 ('FEDRA', 1220),
 ('Luciérnagas', 2639),
 ('MaCdonia', 9125),
 ('Carroñeros', 8841),
 ('Pythonia', 1899),
 ('Luciérnagas', 133),
 ('JavaScriptia', 8448),
 ('Luciérnagas', 9087),
 ('JavaScriptia', 2117),
 ('Luciérnagas', 5495),
 ('JavaScriptia', 6962),
 ('Luciérnagas', 2917),
 ('MaCdonia', 9961),
 ('Carroñeros', 7208),
 ('Carroñeros', 2672),
 ('JavaScriptia', 5616),
 ('MaCdonia', 532),
 ('FEDRA', 9549),
 ('Carroñeros', 1733),
 ('Luciérnagas', 323),
 ('Pythonia', 2117),
 ('Pythonia', 7393),
 ('JavaScriptia', 8167),
 ('FEDRA', 9382),
 ('MaCdonia', 5616)]

### Consulta 8 (6 pts)
**Nombre del Banco, cantidad de usuarios pertenecientes al banco y saldo total del banco.**

###### Es decir, todos los bancos que ocupan los habitantes de cada faccion

Ejemplo:

|Banco | Cantidad|Saldo|
|----|---|---|
|FedraBank S.A |321|20000|
|Banco Pedrito Pascal SpA  |21|200| 
| Banco De Chile|213|56000|
| ... | ... |

**Solución:** Usamos un INNER JOIN para 'intersectar' la información de la tabla personas y cuentas_bancarias, luego sumamos los saldos con SUM() y contamos la cantidad de usuarios con COUNT().

In [200]:
%%sql

/* Reemplace estas queries por las que resuelven las consultas pedidas. */

SELECT
    cuentas_bancarias.banco AS "Banco",
    COUNT(*) AS "Cantidad",
    SUM(cuentas_bancarias.saldo) AS "Saldo"
FROM 
    cuentas_bancarias INNER JOIN personas ON cuentas_bancarias.id_persona = personas.id
GROUP BY (cuentas_bancarias.banco)

 * postgresql://postgres:***@localhost:5432/tarea1
10 rows affected.


[('Banco Consorcio S.A', 10, 61482),
 ('Banco de la Nación S.A', 11, 62308),
 ('Banco Santander SpA', 11, 39473),
 ('Banco Pythonia S.A', 9, 40842),
 ('Banco BCI SpA', 12, 65400),
 ('FedraBank S.A', 16, 72867),
 ('Banco BBVA', 14, 69316),
 ('Banco Estado SpA', 14, 73313),
 ('Banco Pedrito Pascal SpA', 13, 77144),
 ('Banco Scotiabank', 10, 40660)]

### Consulta 9 (8 pts)
**Nombre de la Facción y el Banco más utilizado por sus habitantes**(Si hay empate, que se muestre solo uno.)

Ejemplo:
   
|Facción | Banco |
| - | - |
| FEDRA | FedraBank S.A |
| Luciérnagas | Banco Pedrito Pascal SpA |
| ... | ... |

**Solución:** _Escriba aquí la descripción de su solución._

In [ ]:
%%sql

/* Reemplace estas queries por las que resuelven las consultas pedidas. */

%%sql
SELECT
    * 
FROM 
    *

### Consulta 10 (6 pts)
**ID, nombre y segundo apellido de las personas traidoras**(Es decir, si la persona pertenece a dos o más facciones.)

Ejemplo:
   
|ID | Nombre | Apellido | 
|-|-------|-|
| 1 | Pedro | Pascal |
| 2 | Bella | Ramsey |
| ... | ... | ... | 

**Solución:** _Escriba aquí la descripción de su solución._

In [ ]:
%%sql

/* Reemplace estas queries por las que resuelven las consultas pedidas. */

%%sql
SELECT
    * 
FROM 
    *

### Consulta BONUS (6 pts.)

**Nombre y Apellidos de las personas casadas**(solo se muestra una única vez la pareja)

Ejemplo:
   
|Nombre persona 1 | Apellidos persona 1 | Nombre persona 2 | Apellidos persona 2 |
|-|-------|-|-|
| Felipe | Martinez | Raul | Cuello |
| Rodrigo | Munita | Juan | Mira |
| Nicole | Rojel | Benjamin | Campos |
| ... | ... | ... | ... |

**Solución:** _Escriba aquí la descripción de su solución._

In [ ]:
%%sql

/* Reemplace estas queries por las que resuelven las consultas pedidas. */

%%sql
SELECT
    * 
FROM 
    *